### PR TITLE
Separate shadow pages from wal records and rework wal to use serializer

### PR DESCRIPTION
**Pull request by ray6080**

### CODE REVIEW

**Review by benjaminwinger — APPROVED** [review]

File: src/include/common/cast.h [hunk 1]
```diff
@@ -8,11 +10,15 @@ namespace common {
...
-    if constexpr (std::is_pointer<FROM>()) {
-        KU_ASSERT(newVal != nullptr);
+    try {
+        TO newVal = dynamic_cast<TO>(old);
+        if constexpr (std::is_pointer<FROM>()) {
+            KU_ASSERT(newVal != nullptr);
+        }
+        return newVal;
+    } catch (std::bad_cast& e) {
+        KU_ASSERT(false);
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/common/cast.h:20]: It's just that it's easier to debug with ku_assert than catching the `std::bad_cast` exception manually.

File: src/common/serializer/buffered_file.cpp [hunk 2]
```diff
@@ -58,9 +63,17 @@ void BufferedFileReader::read(uint8_t* data, uint64_t size) {
     }
 }
 
+bool BufferedFileReader::finished() {
+    return bufferOffset >= bufferSize && fileSize <= fileOffset;
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/common/serializer/buffered_file.cpp:67]: `bufferSize = std::min(fileSize - fileOffset, BUFFER_SIZE);` `bufferSize` is bounded by the actual fileSize. so it won't read empty data at the end of the last page if it's partially filled.

[the remark this answers is not in the document]
Comment (ray6080) [reply src/common/serializer/buffered_file.cpp:67]: Ensuring wal is not corrupted is another thing we need to handle. I feel we might need to do checksum for records.

File: src/storage/wal_replayer.cpp [hunk 3]
```diff
@@ -95,17 +108,18 @@ void WALReplayer::replayWALRecord(WALRecord& walRecord,
         replayAddPropertyRecord(walRecord);
     } break;
     default:
-        throw RuntimeException(
-            "Unrecognized WAL record type inside WALReplayer::replay. recordType: " +
-            walRecordTypeToString(walRecord.recordType));
+        throw RuntimeException("Unrecognized WAL record type inside WALReplayer::replay. type: " +
+                               walRecordTypeToString(walRecord.type));
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/storage/wal_replayer.cpp:112]: Yeah. nice catch here. I cleaned up these toString functions. should just put `KU_UNREACHABLE` here.

Removed `CREATE_REL_TABLE_GROUP_RECORD`. not used for now.

File: src/storage/wal/wal_record.cpp [hunk 4]
```diff
@@ -1,10 +1,184 @@
...
+        walRecord = PageUpdateOrInsertRecord::deserialize(deserializer);
+    } break;
+    case WALRecordType::COMMIT_RECORD: {
+        walRecord = CommitRecord::deserialize(deserializer);
+    } break;
+    default: {
+        KU_UNREACHABLE;
+    }
+    }
+    walRecord->type = type;
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/storage/wal/wal_record.cpp:54]: Not really. deserializer uses default constructor which doesn't take in `type`. though a better way is to remove default constructor. I will do that in another pr later.

File: test/storage/node_insertion_deletion_test.cpp [hunk 5]
```diff
@@ -135,5 +135,5 @@ TEST_F(NodeInsertionDeletionTests, TruncatedWalTest) {
         walFileInfo->truncate(BufferPoolConstants::PAGE_4KB_SIZE);
     }
     // Re-open database
-    database = std::make_unique<Database>(databasePath);
+    EXPECT_THROW(database = std::make_unique<Database>(databasePath), RuntimeException);
```
[the remark this answers is not in the document]
Comment (ray6080) [reply test/storage/node_insertion_deletion_test.cpp:138]: I will keep it for now as a test on malformed wal file. we should add more tests on malformed wal, and rework this later.

**Pull request merged**